### PR TITLE
Implicit type inclusion changes

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Implements automatic type inclusion per our discussion today

### CODE REVIEW

File: src/compiler/commandLineParser.ts [hunk 1]
```diff
@@ -337,8 +337,12 @@ namespace ts {
             }
         },
         {
-            name: "typesRoot",
-            type: "string"
+            name: "typeRoots",
+            type: "list",
+            element: {
+                name: "typeRoots",
+                type: "string"
```
Comment (mhegazy) [inline src/compiler/commandLineParser.ts]: isFilePath: true

File: src/compiler/program.ts [hunk 2]
```diff
@@ -197,39 +192,36 @@ namespace ts {
...
-                const primaryPath = combinePaths(rootDir, searchPath);
-                if (traceEnabled) {
-                    trace(host, Diagnostics.Resolving_with_primary_search_path_0, primaryPath);
-                }
-                const candidate = combinePaths(primaryPath, typeReferenceDirectiveName);
+        if (typeRoots.length) {
+            if (traceEnabled) {
+                trace(host, Diagnostics.Resolving_with_primary_search_path_0, typeRoots.join(', '));
+            }
+            const primarySearchPaths = options.typeRoots || defaultTypeRoots;
```
Comment (mhegazy) [inline src/compiler/program.ts]: is not this just `typeRoots`?

File: src/compiler/types.ts [hunk 3]
```diff
@@ -2562,7 +2562,8 @@ namespace ts {
         target?: ScriptTarget;
         traceResolution?: boolean;
         types?: string[];
-        /* @internal */ typesRoot?: string;
+        // Paths used to used to compute primary types search locations
```
Comment (mhegazy) [inline src/compiler/types.ts]: jsdoc

File: src/server/editorServices.ts [hunk 4]
```diff
@@ -315,6 +315,10 @@ namespace ts.server {
             return this.host.directoryExists(path);
         }
 
+        getDirectories(path: string): string[] {
+            return this.host.getDirectories ? this.host.getDirectories(path) : [];
```
Comment (mhegazy) [inline src/server/editorServices.ts]: why would the host not have this function? it is on sys, and we are the only implementers of this API so far.

File: src/services/services.ts [hunk 5]
```diff
@@ -2139,7 +2141,7 @@ namespace ts {
         const getCanonicalFileName = createGetCanonicalFileName(!!useCaseSensitiveFileNames);
 
         function getKeyForCompilationSettings(settings: CompilerOptions): DocumentRegistryBucketKey {
-            return <DocumentRegistryBucketKey>`_${settings.target}|${settings.module}|${settings.noResolve}|${settings.jsx}|${settings.allowJs}|${settings.baseUrl}|${settings.typesRoot}|${settings.typesSearchPaths}|${JSON.stringify(settings.rootDirs)}|${JSON.stringify(settings.paths)}`;
+            return <DocumentRegistryBucketKey>`_${settings.target}|${settings.module}|${settings.noResolve}|${settings.jsx}|${settings.allowJs}|${settings.baseUrl}|${settings.typeRoots}|${JSON.stringify(settings.rootDirs)}|${JSON.stringify(settings.paths)}`;
```
Comment (mhegazy) [inline src/services/services.ts]: `JSON.stringify(settings.typeRoots)`

File: src/compiler/program.ts [hunk 6]
```diff
@@ -1034,21 +1010,34 @@ namespace ts {
...
 
-        // or load all types from the automatic type import fields
-        if (host && host.getDefaultTypeDirectiveNames) {
-            const commonRoot = computeCommonSourceDirectoryOfFilenames(rootFiles, host.getCurrentDirectory(), f => host.getCanonicalFileName(f));
-            if (commonRoot) {
-                return host.getDefaultTypeDirectiveNames(commonRoot);
+        // Walk the primary type lookup locations
+        let result: string[] = [];
+        if (host.directoryExists && host.getDirectories) {
+            for (const root of options.typeRoots || defaultTypeRoots) {
```
Comment (mhegazy) [inline src/compiler/program.ts]: should not we make this absolute? defaultTypeRoots is relative.. correct?

Comment (mhegazy) [inline src/compiler/program.ts]: what about walking up to find the first node_modules? we are not doing this any more?

Comment (RyanCavanaugh) [inline src/compiler/program.ts]: We agreed not to do the walking behavior due to it possibly including undesired folders

**Comment by RyanCavanaugh** [comment]
:bell:

File: src/compiler/program.ts [hunk 7]
```diff
@@ -197,39 +192,36 @@ namespace ts {
             traceEnabled
         };
 
-        // use typesRoot and fallback to directory that contains tsconfig or current directory if typesRoot is not set
-        const rootDir = options.typesRoot || (options.configFilePath ? getDirectoryPath(options.configFilePath) : (host.getCurrentDirectory && host.getCurrentDirectory()));
-
+        const typeRoots = options.typeRoots || defaultTypeRoots;
```
Comment (mhegazy) [inline src/compiler/program.ts]: we should make `defaultTypeRoots` absolute as well.

**Comment by mhegazy** [comment]
We should add a test to `reuseProgramStructure.ts`

File: src/compiler/program.ts [hunk 8]
```diff
@@ -1212,10 +1205,9 @@ namespace ts {
                 (oldOptions.jsx !== options.jsx) ||
                 (oldOptions.allowJs !== options.allowJs) ||
                 (oldOptions.rootDir !== options.rootDir) ||
-                (oldOptions.typesSearchPaths !== options.typesSearchPaths) ||
                 (oldOptions.configFilePath !== options.configFilePath) ||
                 (oldOptions.baseUrl !== options.baseUrl) ||
-                (oldOptions.typesRoot !== options.typesRoot) ||
+                (oldOptions.typeRoots !== options.typeRoots) ||
```
Comment (mhegazy) [inline src/compiler/program.ts]: `arrayIsEqualTo(oldOptions.typeRoots !== options.typeRoots)`

File: src/compiler/program.ts [hunk 9]
```diff
@@ -178,6 +178,11 @@ namespace ts {
 
     const typeReferenceExtensions = [".d.ts"];
 
+    function getEffectiveTypeRoots(options: CompilerOptions, host: ModuleResolutionHost) {
+        return options.typeRoots ||
+            defaultTypeRoots.map(d => combinePaths(options.configFilePath ? getDirectoryPath(options.configFilePath) : host.getCurrentDirectory(), d));
```
Comment (mhegazy) [inline src/compiler/program.ts]: ts.map(defaultTypeRoots

**Comment by mhegazy** [comment]
one comment. other than that 👍

**Comment by mhegazy** [comment]
getInferredTypesRoot seems to be big enough, we should add some unit tests for it, just to make sure we captured all the interesting scenarios.

this does not have to gate this change though.

**Pull request merged**